dartagnan/src/main/java/com/dat3m/dartagnan/parsers/program/visitors/VisitorLitmusAArch64.java

```diff
@@ -105,163 +143,308 @@ public Object visitInstructionRow(LitmusAArch64Parser.InstructionRowContext ctx)
     }
 
     @Override
-    public Object visitMov(LitmusAArch64Parser.MovContext ctx) {
-        Register register = programBuilder.getOrNewRegister(mainThread, ctx.rD, archType);
-        Expression expr = ctx.expr32() != null ? (Expression)ctx.expr32().accept(this) : (Expression)ctx.expr64().accept(this);
-        return programBuilder.addChild(mainThread, EventFactory.newLocal(register, expr));
+    public Object visitMov(MovContext ctx) {
+        Register register = parseRegister64(ctx.r32, ctx.r64);
+        Expression expr = parseExpression(ctx.expr32(), ctx.expr64());
+        return add(EventFactory.newLocal(register, expressions.makeCast(expr, register.getType())));
     }
 
     @Override
-    public Object visitCmp(LitmusAArch64Parser.CmpContext ctx) {
-        Register register = programBuilder.getOrNewRegister(mainThread, ctx.rD, archType);
-        Expression expr = ctx.expr32() != null ? (Expression)ctx.expr32().accept(this) : (Expression)ctx.expr64().accept(this);
+    public Object visitCmp(CmpContext ctx) {
+        Register register = parseRegister64(ctx.r32, ctx.r64);
+        Expression expr = parseExpression(ctx.expr32(), ctx.expr64());
         lastCmpInstructionPerThread.put(mainThread, new CmpInstruction(register, expr));
```

Comment:
Other instructions have a cast here. Why not this one? Does `cmp` guarantees both operands have the same type?

Reply:
Currently, all litmus tests with 32-bit CMP instructions also work when treated as 64-bit comparisons:
```
litmus/AARCH64/EXTRA/ARMARM00.litmus
litmus/AARCH64/EXTRA/ARMARM01.litmus
litmus/AARCH64/EXTRA/ARMARM02.litmus
litmus/AARCH64/EXTRA/ARMARM04.litmus
litmus/AARCH64/EXTRA/ARMARM04+BIS.litmus
litmus/AARCH64/EXTRA/ARMARM05.litmus
litmus/AARCH64/HAND/ARMARM00.litmus
litmus/AARCH64/HAND/ARMARM01.litmus
litmus/AARCH64/HAND/ARMARM02.litmus
litmus/AARCH64/HAND/ARMARM04.litmus
litmus/AARCH64/HAND/ARMARM04+BIS.litmus
litmus/AARCH64/HAND/ARMARM05.litmus
litmus/AARCH64/HAND/ISA2+LA.litmus
litmus/AARCH64/HAND/WRC+poap+addrpa+L.litmus
litmus/AARCH64/HAND/WRC+realdata+realctrlisb.litmus
```